cli/cook/querying.py

```diff
@@ -237,22 +239,30 @@ def __get_latest_instance(job):
     raise Exception(f'Job {job["uuid"]} currently has no instances.')
 
 
-def query_unique_and_run(clusters, entity_ref, command_fn):
+def query_unique_and_run(clusters, entity_ref, command_fn, wait=False):
     """Calls query_unique and then calls the given command_fn on the resulting job instance"""
-    query_result = query_unique(clusters, entity_ref)
-    if query_result['type'] == Types.JOB:
-        job = query_result['data']
-        instance = __get_latest_instance(job)
-        directory = mesos.retrieve_instance_sandbox_directory(instance, job)
-        command_fn(instance, directory)
-    elif query_result['type'] == Types.INSTANCE:
-        instance, job = query_result['data']
-        directory = mesos.retrieve_instance_sandbox_directory(instance, job)
-        command_fn(instance, directory)
+
+    def query_unique_and_run():
+        query_result = query_unique(clusters, entity_ref)
+        if query_result['type'] == Types.JOB:
+            job = query_result['data']
+            instance = __get_latest_instance(job)
+            directory = mesos.retrieve_instance_sandbox_directory(instance, job)
+            command_fn(instance, directory)
+        elif query_result['type'] == Types.INSTANCE:
+            instance, job = query_result['data']
+            directory = mesos.retrieve_instance_sandbox_directory(instance, job)
+            command_fn(instance, directory)
+        else:
+            # This should not happen, because query_unique should
+            # only return a map with type "job" or type "instance"
+            raise Exception(f'Encountered error when querying for {entity_ref}.')
+
+    if wait:
+        r = tenacity.Retrying(wait=tenacity.wait_fixed(5))
+        r.call(query_unique_and_run)
```

Comment:
Does this eventually give up?  Should we consider retrying only for specific exceptions?

Reply:
This never gives up. I introduced `CookRetriableException` and switched to using it in the specific spots where we want retrying.

Reply:
I like this change. Now we won't get into an infinite retry loop when there's some unexpected error.

Reply:
How about if we add a separate flag (e.g. `--wait-for '5 minutes'`) for the eventually giving up case, and leave the `--wait` as an indefinite wait? It's definitely true that those exceptions may not get resolved. As a user, though, I may want this to block until they do get resolved. For example, I submit a job, and it may take a long time for it to start, and then once it starts it may take a long time for my file to appear, but I would prefer to just wait until those things are true without having to decide if I should wait 30 minutes or 90 minutes or longer.

Reply:
@sradack What are your thoughts on this?

Reply:
@sradack ping

Reply:
@sradack I added a 1-day timeout on the `--wait`, and it re-raises the last exception